fix(past-answers) rework attempt count page, solving N+1 and other minor issues

adi-herwana-nus opened the pull request:
- removed old #question_answer_details and #all_answers apis from statistics answer controller
- created new #show and #all_answers apis on statistics answer controller with tests
- removed past attempts statistics page
- migrated marks/question and attempt count pages to new APIs
- reimplemented AllAttemptsDisplay component to fetch/cache individual answers on slider release
- display attempt index on active attempt slider node
- not started submissions no longer link on attempt count page
- display attempt counts also for submissions that are still attempting

cysjonathan commented on client/app/bundles/course/assessment/pages/AssessmentStatistics/AnswerDisplay/AllAttemptsDisplay.tsx:
```diff
@@ -63,63 +110,113 @@ const AllAttemptsDisplay: FC<Props> = (props) => {
-        }
-        question={question}
-      />
+      <div className={sliderIndex === displayedIndex ? '' : 'opacity-60'}>
```
nit: consider `sliderIndex !== displayedIndex && 'opacity-60'`  
but don't feel particularly strongly about it either way

adi-herwana-nus replied:
I tried that and it doesn't work on native divs. Perhaps it only works on React components?

cysjonathan commented on client/app/bundles/course/assessment/pages/AssessmentStatistics/StudentAttemptCountTable.tsx:
```diff
@@ -22,20 +26,54 @@ import { getAssessmentStatistics } from './selectors';
+  workflowState?: WorkflowState | typeof workflowStates.Unstarted;
+}
+
+const StudentSubmissionChip: FC<StudentSubmissionChipProps> = (props) => {
```
Duplicated code in `StudentMarksPerQuestionTable.tsx`, so should extract this as a common component and use sitewide.  
Ideally check if we can also replace chip in `/submissions` page `client/app/bundles/course/assessment/submission/pages/SubmissionsIndex/SubmissionsTableRow.jsx` L190

cysjonathan commented on client/app/bundles/course/assessment/pages/AssessmentStatistics/StudentAttemptCountTable.tsx:
```diff
@@ -22,20 +26,54 @@ import { getAssessmentStatistics } from './selectors';
+const StudentSubmissionChip: FC<StudentSubmissionChipProps> = (props) => {
+  const { className, linkTo, workflowState } = props;
+  const chip = (
+    <Chip
```
Consider instead using `component={Link}` and `to={linkTo}` attributes instead.  
https://mui.com/material-ui/api/chip/#props

cysjonathan commented on config/routes.rb:
```diff
@@ -437,13 +437,13 @@
         get 'course/progression' => 'aggregate#course_progression'
         get 'course/performance' => 'aggregate#course_performance'
-        get 'submission_question/:id' => 'answers#all_answers'
+        get 'submission/:submission_id/question/:question_id' => 'answers#all_answers'
```
nit: similar to `answers/:id`, idiomatic pattern should be `submissions/:submission_id/questions/:question_id`

adi-herwana-nus commented on client/app/bundles/course/assessment/pages/AssessmentStatistics/AnswerDisplay/AllAttemptsDisplay.tsx:
```diff
@@ -63,63 +110,113 @@ const AllAttemptsDisplay: FC<Props> = (props) => {
+      setDisplayedIndex(newIndex);
+    } else {
+      tryFetchAnswerById(allAnswers[newIndex].id).finally(() => {
+        setSliderIndex(newIndex);
+        setDisplayedIndex(newIndex);
```
Done (wrapped the index updates in a try/finally block to maintain sync between UI and page state)

adi-herwana-nus commented on client/app/bundles/course/assessment/pages/AssessmentStatistics/AnswerDisplay/AllAttemptsDisplay.tsx:
```diff
@@ -63,63 +110,113 @@ const AllAttemptsDisplay: FC<Props> = (props) => {
+      {renderQuestionComponent()}
       {answerSubmittedTimes.length > 1 && (
-        <div className="w-[calc(100%_-_17rem)] mx-auto">
+        <div className="w-[calc(100%_-_17rem)] mx-auto pt-8">
```
I'll leave that as is for now, it seems to handle small screens better

adi-herwana-nus commented on client/app/bundles/course/assessment/pages/AssessmentStatistics/AnswerDisplay/AllAttemptsDisplay.tsx:
```diff
@@ -63,63 +110,113 @@ const AllAttemptsDisplay: FC<Props> = (props) => {
-              setDisplayedIndex(Array.isArray(value) ? value[0] : value);
+            onChange={(event, value) => {
+              const newIndex = Array.isArray(value) ? value[0] : value;
+              if (event.type.includes('mouse')) {
```
Updated to listen to 'touch' events as well.

I tried filtering on 'pointer' and it didn't work for either

adi-herwana-nus commented on client/app/bundles/course/assessment/pages/AssessmentStatistics/AnswerDisplay/AllAttemptsDisplay.tsx:
```diff
@@ -63,63 +110,113 @@ const AllAttemptsDisplay: FC<Props> = (props) => {
+  };
+
+  const renderAnswerDetailsComponent = (): JSX.Element | null => {
+    if (answerStatus === 'completed') {
+      return (
+        <>
+          <Typography variant="body1">
+            {t(translations.pastAnswerTitle, {
+              submittedAt: formatLongDateTime(answerDetails!.createdAt),
+            })}
+          </Typography>
+          <AnswerDetails answer={answerDetails!} question={question!} />
+        </>
+      );
+    }
+    if (answerStatus === 'errored') {
+      return (
+        <Alert severity="error">{t(messagesTranslations.fetchingError)}</Alert>
+      );
+    }
+    if (answerStatus === 'pending') {
+      return <LoadingIndicator />;
+    }
+    return null;
```
I don't think that fits neatly, since the completed status needs the answer details while the others don't by design

adi-herwana-nus commented:
Pushed with the following changes:
- Updated to latest master
- removed various dead items (`MAX_ANSWERS_COUNT`, `isAnswerDisplayed`)
- refactored all the submission status chips into a single component (primarily based on the one from `SubmissionTableRow`, with features from the other instances integrated)
- Reworded apis to use plural forms `.../answers/:id`,  `.../submissions/:submission_id/questions/:question_id`,

cysjonathan reviewed (changes requested):
Two nits then good-to-go:
1. Attempt count not reflective of number of answers (e.g. shows 1 attempt count, but when modal opens, I see two answers). I think we can simply show submitted answers (aligns with the behaviour for "past answers" in the submission page.  
Need to filter that out in the backend.  
2. In Marks-per-Question tab, when modal is opened, can we recycle `SubmissionWorkflowState` to replace with the `[Go to Answer Page]` link?

adi-herwana-nus commented:
> Two nits then good-to-go:
> 
> 1. Attempt count not reflective of number of answers (e.g. shows 1 attempt count, but when modal opens, I see two answers). I think we can simply show submitted answers (aligns with the behaviour for "past answers" in the submission page.
>    Need to filter that out in the backend.
> 2. In Marks-per-Question tab, when modal is opened, can we recycle `SubmissionWorkflowState` to replace with the `[Go to Answer Page]` link?

Done and done.
